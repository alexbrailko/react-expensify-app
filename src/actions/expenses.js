import uuid from 'uuid';
import database from '../firebase/firebase';

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData; //destructure from expense data
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
    return (dispatch) => { //return function that will do async work. dispatch gets passed to this function by the redux library
        return database.ref(`expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
    return (dispatch) => { //return function which will have access to dispatch
        return database.ref(`expenses/${id}`).update(updates).then(() => {
            dispatch(editExpense(id, updates));
        });
    }
};

// SET EXPENSES
// will manipulate redux store
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//async action. will fetch the data
export const startSetExpenses = () => { 

    return (dispatch) => { //return function which will have access to dispatch
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({ //push expenses on an array
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    }
} 