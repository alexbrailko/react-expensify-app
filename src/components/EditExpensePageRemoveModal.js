import React from 'react';
import ReactDOM from 'react-dom';
import Modal from 'react-modal';

const customStyles = {
    content : {
      top                   : '50%',
      left                  : '50%',
      right                 : 'auto',
      bottom                : 'auto',
      marginRight           : '-50%',
      transform             : 'translate(-50%, -50%)'
    }
};

const RemoveModal = (props) => (
    <Modal
    isOpen={props.modalIsOpen}
    onRequestClose={props.handleClearSelectedOption}
    contentLabel="Are you sure you want to delete?"
    style={customStyles}
    >
        <h3>Are you sure you want to delete?</h3>
        <button className="yes" onClick={props.removeExpense}>yes</button>
        <button className="no" onClick={props.closeModal}>no</button>
    </Modal>
);

export default RemoveModal;