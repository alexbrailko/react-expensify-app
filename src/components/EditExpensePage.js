import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import Modal from 'react-modal';

const modalStyles = {
    content : {
      top                   : '50%',
      left                  : '50%',
      right                 : 'auto',
      bottom                : 'auto',
      marginRight           : '-50%',
      transform             : 'translate(-50%, -50%)'
    }
};

export class EditExpensePage extends React.Component {
    state = {
        modalIsOpen: false
    };

    closeModal = () => {
        this.setState(() => ({ modalIsOpen: false }));
    };

    openModal = () => {
        this.setState(() => ({ modalIsOpen: true }));
    };

    onSubmit = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };


    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary removeButton" onClick={this.openModal}>Remove</button>
                </div>
                <Modal
                isOpen={this.state.modalIsOpen}
                onRequestClose={this.closeModal}
                contentLabel="Are you sure you want to delete?"
                style={modalStyles}
                >
                    <h3>Are you sure you want to delete?</h3>
                    <button className="yes" onClick={this.onRemove}>yes</button>
                    <button className="no" onClick={this.closeModal}>no</button>
                </Modal>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => { //give the component the currect expense object
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  } 
}; //pass new props

const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)), 
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
