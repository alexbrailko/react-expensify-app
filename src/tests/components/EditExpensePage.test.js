import React from 'react';
import { shallow } from 'enzyme';
import { findDOMNode } from 'react-dom';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';
import Modal from 'react-modal';

let startEditExpense,  startRemoveExpense, history, wrapper;

beforeEach(() => {
    startEditExpense = jest.fn();
    startRemoveExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditExpensePage 
        startEditExpense={startEditExpense} 
        startRemoveExpense={startRemoveExpense} 
        history={history} 
        expense={expenses[2]}
        />
    );
});

test('should render EditExpensePage correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle startEditExpense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2] );
});

test('should handle startRemoveExpense', () => {
    wrapper.find('button.yes').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[2].id });
});

test('should render modal', () => {
    const modal = wrapper.find('Modal');
    wrapper.find('button.removeButton').simulate('click');
    expect(modal).toHaveLength(1); 
    expect(wrapper.state('modalIsOpen')).toEqual(true);
});
 