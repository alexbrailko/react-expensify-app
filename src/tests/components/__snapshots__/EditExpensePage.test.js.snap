// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "credit card",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary removeButton"
      onClick={[Function]}
    >
      Remove
    </button>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Are you sure you want to delete?"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
        },
      }
    }
  >
    <h3>
      Are you sure you want to delete?
    </h3>
    <button
      className="yes"
      onClick={[Function]}
    >
      yes
    </button>
    <button
      className="no"
      onClick={[Function]}
    >
      no
    </button>
  </Modal>
</div>
`;
